### DISCUSSION TITLE
interface implementation variables - "impl vars"

### DISCUSSION THREAD

**Discussion by JeffGillin** [dotnet/csharplang#6421] [Language Ideas]
Hi all,
    I recently came up with an idea for the Rust language that I want to share here as well.
    The concept revolves around increasing separation of concerns and cohesiveness. We're blocking other methods from accessing our "implementation variables" (impl vars) - private variables that the others don't need to know about.
    The idea is simple. All of the methods required to be implemented for a class are adding to a "block" (an impl block like in rust). Once declared in the impl block, it's a "private class member", just like if it was defined anywhere in class outside of methods, however, it is only accessible to the methods within that block (other trait implementation methods), or any other "helper methods" defined in that same block.
    While there could be some implementation challenges (no pun  intended), I think the idea here is language changing. Very much help understand exactly what a variables purpose is, and where it gets used.
    I'm probably going to submit to Java and a few other languages as well (don't let them be first).
    I'm actually fully on Rust these days, but I always liked C# when I was using it. Very "usable" language.
    Thoughts?
Jeff

**Comment by HaloFour** [comment]
C# already has syntax to privately implement the members of an interface:

```cs
public class Foo : IDisposable {
    void IDisposable.Dispose() { }

    public void M() {
        this.Dispose(); // compiler error
    }
}
```

**Reply by JeffGillin** [reply]
I'm not sure what you are showing above exactly. (Edit: Actually, I see what you're showing now, but it isn't related to my idea actually). The idea revolves around "variables", and limiting their scope.
Not sure if you have seen the "impl" block in Rust.
But, imagine in C# as the following

```
class Foo : ICalcTotals {
     int total = 0;  
  
    impl ICalcTotals{
         int partialTotal1 = 0;    // Hi, I'm an "impl var"
         int partialTotal2 = 0    // me too
         
      pubilc void calcSomething(){
                // calling methods outside of impl block is OK, they just can't access the vars defined in this impl block
                // You can pass it as a parameter tho of course, but they won't be able to change the value unless the method the parameter is passed to is defined in this impl block.
                 partialTotal1 += some_class.some_method();   
                 partialTotal2 += calc_partialTotal2();   
        }

        pubiic void update_full_total(){
               this.total = partialTotal1 + partialTotal2;
       }
   }
}
```

This is the basic syntax. I can't overstate how much this would have helped on many projects. So much so, that when I look at it now, I think how could this not be in the languages already.

**Reply by HaloFour** [reply]
I see, so it's a method of grouping the members together.  I did not get that from your original post.

There is a very limited proposal to consider doing this with properties, so that they can have explicit backing fields but without making them available directly to the rest of the class here: #133

There was a previous discussion about allowing members to be declared within scope blocks on the Roslyn repository here: https://github.com/dotnet/roslyn/issues/12361

You could achieve something like this today by writing an analyzer.  That analyzer could analyze which fields are used by which methods and, based on some additional metadata likes attributes, comments or regions, could emit an error if the use isn't explicitly declared.

**Reply by JeffGillin** [reply]
Not about grouping tho. About further HIDING private vars - that's the important part. They will be "out of scope" for those outside this impl block (with the possible exception for variables passed from inside the block to outside (that would be a consideration - might be able to pass an object and let fields be changed outside this block - perhaps).

When I implement MY trait (cue Ross - sandwich episode), I don't want others getting confused and thinking my variables can be touched by them just because on the same class.

**Reply by HaloFour** [reply]
> Not about grouping tho. About further HIDING private vars - that's the important part.

It is about grouping members into a scope.  Those "private vars" are fields, which are members just like methods are.  There's no other concept of class-level variables in C#.

IMO, if the class is so large that you need to create boundaries between the members within that class that would be an indicator that the class is much too large and trying to cover too many concerns and should be refactored into smaller types.

**Reply by JeffGillin** [reply]
There's truth to what you say regarding "small classes", but this is another separate concern. Classes implement several interfaces, and non interface methods, and we don't want those touching the private variables of our interface implementation (ones that were private class members, but now tucked away from view of most)

**Comment by svick** [comment]
The goal is different, but this sounds quite similar to "file private visibility" (https://github.com/dotnet/csharplang/issues/5529), except that with that, you would have to declare each "block" in a separate file.

**Reply by HaloFour** [reply]
Is that still being considered for members within the class?  If so that would meet the need and also provide a way to organize the code:

```cs
// MyClass.cs
public partial class MyClass {

    public int DoSomething() {
        _myPrivateState = 0; // compiler error
        return this.M(); // fine
    }
}

// MyClass.IMyInterface.cs (or any other name you want)
public partial class MyClass : IMyInterface {
    file private int _myPrivateState;

    private int M() {
        return ++_myPrivateState;
    }
}
```

**Reply by 333fred** [reply]
> Is that still being considered for members within the class?

It was designed with future expansion in mind, though that expansion is not currently planned.

**Comment by TahirAhmadov** [comment]
I think the problem with creating nested "scopes" within a type introduces other questions. For example, what about `protected` and `internal` members inside those scopes, what are their visibility rules? What about `virtual` members? Etc.

**Reply by CyrusNajmabadi** [reply]
Presumably all 'variables' would be entirely private. It would be better to think of them as 'locals' versus any sort of actual member.

**Reply by TahirAhmadov** [reply]
In his example, the interface implementing methods are `public`. I guess we could devise a set of rules: you can have public members; those are just like other public members outside of "scopes"; all other members have to be private; the public members cannot be virtual; etc.

In short, I think this is way too complicated to be of any benefit.

**Reply by CyrusNajmabadi** [reply]
> In his example

> > I'm not sure what you are showing above exactly. (Edit: Actually, I see what you're showing now, but it isn't related to my idea actually). The idea revolves around "variables", and limiting their scope.

> In short, I think this is way too complicated to be of any benefit.

Seems beneficial to me.  We're already thinking about doing this already for properties.  Extending to a set of members is definitely a possibility, though we'll likely make the determination when we do the property-locals design.

**Reply by HaloFour** [reply]
> Presumably all 'variables' would be entirely private. It would be better to think of them as 'locals' versus any sort of actual member.

What about methods, though?  Are they "local functions"?  Could fields have accessibility modifiers?  What does a `public` method within one of these scopes mean?  What about a `public` field?  Can names collide between these scopes?

I can see there being some limited utility in being able to hide portions of state from parts of the class where it's possible that accidentally setting fields directly could cause issues.  I think scoping within a property makes more sense as the property becomes the gate keeper to at least ensuring that appropriate logic is applied.  Doing it across arbitrary sets of members feels like attempting to erect artificial boundaries for the sake of mitigating the problem of super-types.  But the developers who would employ such a mechanism are the same developers who should be encouraged to break up those super-types into smaller components.

I'd rather solve problems like this by nested types and delegation, the latter of which would have provide more utility to the language in general.

And this is also a problem that can be solved via analyzers without the need for any language changes.

**Reply by TahirAhmadov** [reply]
@HaloFour good points - plus the metadata/IL question will have to be answered WRT names of these fields.

If I was presented with such a challenge, I would most likely create a nested struct and delegate the interface implementation to that struct.

**Comment by TahirAhmadov** [comment]
@JeffGillin why not just do this:
```cs
class Foo : ICalcTotals {
  public Foo()
  {
    this._calcTotalsImpl = new CalcTotalsImpl(this);
  }

  pubilc void calcSomething() => this._calcTotalsImpl.calcSomething();
  pubiic void update_full_total() => this._calcTotalsImpl.update_full_total();

  int total = 0;  
  CalcTotalsImpl _calcTotalsImpl;
  
  struct CalcTotalsImpl{
    public CalcTotalsImpl(Foo owner) => this._owner = owner;
    Foo _owner;
    int partialTotal1 = 0;    // Hi, I'm an "impl var"
    int partialTotal2 = 0    // me too
         
    pubilc void calcSomething(){
      // calling methods outside of impl block is OK, they just can't access the vars defined in this impl block
      // You can pass it as a parameter tho of course, but they won't be able to change the value unless the method the parameter is passed to is defined in this impl block.
      partialTotal1 += some_class.some_method();   
      partialTotal2 += calc_partialTotal2();
    }

    pubiic void update_full_total(){
      this._owner.total = partialTotal1 + partialTotal2;
    }
  }
}
```
